src/mapgen/mg_decoration.h

```diff
@@ -24,6 +24,7 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include "objdef.h"
 #include "noise.h"
 #include "nodedef.h"
+#include "mapgen/treegen.h"
```

Comment:
You could avoid including this header by using pointers (or references for that matter):

```
namespace treegen {
	struct TreeDef;
};

// .... in DecoLSystem:

treegen::TreeDef *tree_def = nullptr;
```

Reply:
It now uses a shared pointer and `mapgen/treegen.h` does not have to be included anymore.
It's a shared pointer, since the `DecoLSystem` has a clone function. (I haven't looked up how often it will actually be cloned.)

I don't know if this it is better than storing the `TreeDef` inside the struct, but it is forward declared now, as you suggested.